### ISSUE TITLE
ELMo single sentence speed

### ISSUE THREAD

**Issue by kwhumphreys** [allenai/allennlp#1968]
Any recommendations for speeding up `embed_sentence` for runtime use on many single sentences, rather than a large batch? Adding a CUDA 10.0 device has no noticeable impact.

On a p2.xlarge instance, running with CPU only averages just under 0.1sec/sentence when called repeatedly with single sentences, and adding GPU gives no noticeable speedup. (GPU on a large batch gives ~8x speedup.)

Is there a way to load and reuse the GPU for single sentences?

**Comment by shocho3858** [comment]
> Any recommendations for speeding up `embed_sentence` for runtime use on many single sentences, rather than a large batch? Adding a CUDA 10.0 device has no noticeable impact.
> 
> On a p2.xlarge instance, running with CPU only averages just under 0.1sec/sentence when called repeatedly with single sentences, and adding GPU gives no noticeable speedup. (GPU on a large batch gives ~8x speedup.)
> 
> Is there a way to load and reuse the GPU for single sentences?

hi, @kwhumphreys , I wonder how I can use GPU for inference in the following example.

```from allennlp.modules.elmo import Elmo, batch_to_ids

options_file = "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_options.json"
weight_file = "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_weights.hdf5"

elmo = Elmo(options_file, weight_file, 2, dropout=0)

# use batch_to_ids to convert sentences to character ids
sentences = [['First', 'sentence', '.'], ['Another', '.']]
character_ids = batch_to_ids(sentences)

embeddings = elmo(character_ids)

# embeddings['elmo_representations'] is length two list of tensors.
# Each element contains one layer of ELMo representations with shape
# (2, 3, 1024).
#   2    - the batch size
#   3    - the sequence length of the batch
#   1024 - the length of each ELMo vector.0
```

**Comment by kwhumphreys** [comment]
@shocho3858 I'm using `allenllp.commands.elmo.Elmo`, not `allennlp.modules.elmo.Elmo`, and setting `cuda_device` on init here https://github.com/allenai/allennlp/blob/master/allennlp/commands/elmo.py#L147

If you want to use the module, you might be able to use `embeddings = elmo(character_ids).cuda()`, but I think you will have to modify `batch_to_ids` to also use `.cuda()` there.

**Comment by schmmd** [comment]
@kwhumphreys I don't know how you would speed up ELMo performance without batching.  Are you using it in an interactive setting where latency is important?  When I processed a large corpus recently we batched appropriately and parallelized heavily.

**Comment by kwhumphreys** [comment]
Yes, tried a prototype for interactive use, but it was too slow. Then tried adding a Tesla K80 GPU with CUDA 10.0 and it made no difference.
So per-batch initialization is expensive and doesn't use the GPU?

**Comment by kwhumphreys** [comment]
right now 1, but that's obviously not the way to do it :-)

**Comment by kwhumphreys** [comment]
Imagine something like grammar checking and trying to use ELMo to rank acceptability of multiple suggested rewrites. If there are 5 errors in a doc and each error has 5 suggested rewrites, that's a 2.5s delay for ELMo alone (without batching).

**Comment by matt-gardner** [comment]
Ah, so now that sounds like something that you can batch.  Yes?

**Comment by kwhumphreys** [comment]
If batches of say 10-20 sentences are big enough to make a difference?

**Comment by schmmd** [comment]
@kwhumphreys any luck with adding some batching to get a speedup?

**Comment by kwhumphreys** [comment]
@schmmd looks like a 6x speedup from GPU when running small batches

**Comment by guotong1988** [comment]
Thank you. Same question: The eval of ELMo speed.
